### PR TITLE
resolved: Add support for mDNS server

### PR DESCRIPTION
**Pull request by rojkov**
This PR adds minimal implementation of mDNS server to systemd-resolved.

Also it fixes resolving mDNS queries and thus supersedes https://github.com/systemd/systemd/pull/4709.

### CODE REVIEW

**Comment by rojkov** [comment]
If a review is blocked by the CI failure for the s390 arch how one can reproduce it?

**Comment by grawity** [comment]
Doesn't matter, the build log points at Launchpad being broken.

**Comment by martinpitt** [comment]
I retried the s390x test.

**Comment by rojkov** [comment]
The test setup seems to be flaky: can't fetch the sources from time to time if I interpreted the logs correctly.

**Comment by nanonyme** [comment]
@rojkov FWIW looks like DNS probably issue. Can the DNS be made less flaky somehow?

**Comment by martinpitt** [comment]
Sorry, my retry script still used the old git branch URL which doesn't exist any more. Fixed now and retried again.

**Comment by rojkov** [comment]
@martinpitt Thanks! Seems all is OK now.

**Comment by rojkov** [comment]
@poettering ping. Is there a plan for mDNS in resolved? Is this PR relevant?

**Comment by amckinlay** [comment]
Will there be documentation added to `man systemd-resolved.service` about how resolution using mDNS fits in with LLMNR/unicast DNS? Currently, the option `MulticastDNS=` exists in `man systemd.network` but it seem to be a no-op and there is minimal documentation.

**Comment by rojkov** [comment]
@amckinlay IMHO mDNS/LLMNR/DNS have been documented already quite well with their respective RFCs. DNS was designed to be extensible, so the protocols can work together without interference.

The option `MulticastDNS=` lacks implementation indeed. And this PR is exactly about bringing the declared but missing functionality to `systemd-resolved`.

**Comment by keszybz** [comment]
> Is there a plan for mDNS in resolved? Is this PR relevant?

Yes. We're just a bit behind on some reviews.

File: src/resolve/resolved-mdns.c [hunk 1]
```diff
@@ -77,6 +77,9 @@ static int on_mdns_packet(sd_event_source *s, int fd, uint32_t revents, void *us
         if (r <= 0)
                 return r;
 
+        if (manager_our_packet(m, p))
```
Comment (keszybz) [inline src/resolve/resolved-mdns.c:126]: Hm, can you add a comment to the patch message how that'd be possible?

File: src/resolve/resolved-dns-scope.c [hunk 2]
```diff
@@ -549,7 +549,7 @@ static int dns_scope_multicast_membership(DnsScope *s, bool b, struct in_addr in
                         .imr_ifindex = s->link->ifindex,
                 };
 
-                fd = manager_llmnr_ipv4_udp_fd(s->manager);
+                fd = s->protocol == DNS_PROTOCOL_LLMNR ? manager_llmnr_ipv4_udp_fd(s->manager) : manager_mdns_ipv4_fd(s->manager);
```
Comment (keszybz) [inline src/resolve/resolved-dns-scope.c]: Minor thing, but `if` would be more readable then `?:` here.

File: src/resolve/resolved-dns-transaction.c [hunk 3]
```diff
@@ -363,6 +363,9 @@ void dns_transaction_complete(DnsTransaction *t, DnsTransactionState state) {
         SET_FOREACH_MOVE(z, t->notify_zone_items_done, t->notify_zone_items)
                 dns_zone_item_notify(z);
         SWAP_TWO(t->notify_zone_items, t->notify_zone_items_done);
+        if (t->probing) {
+                dns_scope_announce(t->scope);
+        }
```
Comment (keszybz) [inline src/resolve/resolved-dns-transaction.c]: No braces needed.

Comment (keszybz) [reply src/resolve/resolved-dns-transaction.c]: Oh, if it's changed in a later commit, then please ignore.

**Comment by rojkov** [comment]
@keszybz Thank you for review!

* removed unneeded curly braces;
* replaced unreadable `?:` with `if`;
* added a note to NEWS and updated `systemd-resolve`'s man page;
* actually enabled mDNS;
* added a comment about multicast loopback switched on by default and in-process filtering looped back packets (probably makes sense to drop the `manager_our_packet()` from both MDNS and LLMNR code and to switch loopback off);
* rebased and force pushed.

**Comment by keszybz** [comment]
Thanks. I'll try to test this today or tomorrow.

**Comment by keszybz** [comment]
BTW, I haven't forgotten about this, but all the time things which need higher priority keep popping up.

**Comment by keszybz** [comment]
Seems to work nicely.

**Comment by keszybz** [comment]
Thanks!

**Pull request merged**